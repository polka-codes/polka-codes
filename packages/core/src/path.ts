/**
 * Returns the directory portion of a path string.
 * Strips trailing slashes, then takes everything up to the last slash.
 */
export function dirname(path: string): string {
  if (path.length === 0) return '.'
  const isRooted = path[0] === '/'
  // Trim trailing slashes (but leave root "/")
  let end = path.length - 1
  while (end > 0 && path[end] === '/') end--
  const idx = path.lastIndexOf('/', end)
  if (idx < 0) {
    return isRooted ? '/' : '.'
  }
  if (isRooted && idx === 0) {
    return '/'
  }
  return path.slice(0, idx)
}

/**
 * Normalizes a path by resolving "." and ".." segments.
 */
function normalize(path: string): string {
  const isAbsolute = path.startsWith('/')
  const segments = path.split('/').filter(Boolean)
  const stack: string[] = []
  for (const seg of segments) {
    if (seg === '.') continue
    if (seg === '..') {
      if (stack.length && stack[stack.length - 1] !== '..') {
        stack.pop()
      } else if (!isAbsolute) {
        stack.push('..')
      }
    } else {
      stack.push(seg)
    }
  }
  let result = stack.join('/')
  if (!result && !isAbsolute) return '.'
  if (result && path.endsWith('/')) result += '/'
  return (isAbsolute ? '/' : '') + result
}

/**
 * Joins all given path segments, then normalizes the result.
 */
export function join(...parts: string[]): string {
  if (parts.length === 0) return '.'
  let combined = ''
  for (const p of parts) {
    if (typeof p !== 'string') {
      throw new TypeError('Arguments to join must be strings')
    }
    if (p) {
      combined = combined ? `${combined}/${p}` : p
    }
  }
  return normalize(combined)
}
