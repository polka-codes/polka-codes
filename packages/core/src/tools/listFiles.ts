import { z } from 'zod'
import { type FullToolInfoV2, PermissionLevel, type ToolHandler, type ToolInfoV2, ToolResponseType } from '../tool'
import type { FilesystemProvider } from './provider'

export const toolInfo = {
  name: 'list_files',
  description:
    'Request to list files and directories within the specified directory. If recursive is true, it will list all files and directories recursively. If recursive is false or not provided, it will only list the top-level contents. Do not use this tool to confirm the existence of files you may have created, as the user will let you know if the files were created successfully or not.',
  parameters: z.object({
    path: z
      .string()
      .describe('The path of the directory to list contents for (relative to the current working directory)')
      .meta({ usageValue: 'Directory path here' }),
    maxCount: z.coerce
      .number()
      .optional()
      .default(2000)
      .describe('The maximum number of files to list. Default to 2000')
      .meta({ usageValue: 'Maximum number of files to list (optional)' }),
    recursive: z
      .preprocess((val) => {
        if (typeof val === 'string') {
          const lower = val.toLowerCase()
          if (lower === 'false') return false
          if (lower === 'true') return true
        }
        return val
      }, z.boolean().optional().default(true))
      .describe('Whether to list files recursively. Use true for recursive listing, false or omit for top-level only.')
      .meta({ usageValue: 'true or false (optional)' }),
  }),
  examples: [
    {
      description: 'Request to list files',
      parameters: [
        {
          name: 'path',
          value: 'src',
        },
        {
          name: 'maxCount',
          value: '100',
        },
      ],
    },
  ],
  permissionLevel: PermissionLevel.Read,
} as const satisfies ToolInfoV2

export const handler: ToolHandler<typeof toolInfo, FilesystemProvider> = async (provider, args) => {
  if (!provider.listFiles) {
    return {
      type: ToolResponseType.Error,
      message: 'Not possible to list files. Abort.',
    }
  }

  const { path, maxCount, recursive } = toolInfo.parameters.parse(args)
  const [files, limitReached] = await provider.listFiles(path, recursive, maxCount)

  return {
    type: ToolResponseType.Reply,
    message: `<list_files_path>${path}</list_files_path>
<list_files_files>
${files.join('\n')}
</list_files_files>
<list_files_truncated>${limitReached}</list_files_truncated>`,
  }
}

export const isAvailable = (provider: FilesystemProvider): boolean => {
  return !!provider.listFiles
}

export default {
  ...toolInfo,
  handler,
  isAvailable,
} satisfies FullToolInfoV2
