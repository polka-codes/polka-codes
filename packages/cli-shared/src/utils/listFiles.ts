import { promises as fs } from 'node:fs'
import { join, relative, resolve } from 'node:path'
import ignore, { type Ignore } from 'ignore'

/** Default patterns commonly ignored in projects of various languages. */
const DEFAULT_IGNORES = [
  '__pycache__',
  '.DS_Store',
  '.env',
  '.git',
  '.idea',
  '.svn',
  '.temp',
  '.vscode',
  'coverage',
  'dist',
  'node_modules',
  'out',
  'Thumbs.db',
]

/**
 * Reads a `.gitignore` file in `dirPath` (if it exists) and appends its lines
 * to the `basePatterns`. Returns a new array without mutating the original.
 */
async function extendPatterns(basePatterns: string[], dirPath: string): Promise<string[]> {
  try {
    const gitignorePath = join(dirPath, '.gitignore')
    const content = await fs.readFile(gitignorePath, 'utf8')
    const lines = content.split(/\r?\n/).filter(Boolean)
    return [...basePatterns, ...lines]
  } catch {
    // No .gitignore or unreadable
    return basePatterns
  }
}

/** Creates an `ignore` instance from the given patterns. */
function createIgnore(patterns: string[]): Ignore {
  return ignore().add(patterns)
}

/**
 * Lists files under `dirPath` in BFS order, respecting:
 *   - A default set of ignores
 *   - A root .gitignore under `cwd`
 *   - Any .gitignore files in child directories (merged as we go)
 *
 * Returns `[files, limitReached]`:
 *   - `files` is the array of file paths (relative to `cwd`)
 *   - `limitReached` is `true` if `maxCount` was hit, otherwise `false`
 *   - When truncated, adds markers like `path/to/dir/(files omitted)` for truncated directories
 */
export async function listFiles(
  dirPath: string,
  recursive: boolean,
  maxCount: number,
  cwd: string,
  excludeFiles?: string[],
): Promise<[string[], boolean]> {
  // Merge default ignores with root .gitignore and excludeFiles (if found)
  let rootPatterns = [...DEFAULT_IGNORES, ...(excludeFiles || [])]
  try {
    const rootGitignore = await fs.readFile(join(cwd, '.gitignore'), 'utf8')
    const lines = rootGitignore.split(/\r?\n/).filter(Boolean)
    rootPatterns = [...rootPatterns, ...lines]
  } catch {
    // No .gitignore at root or unreadable; ignore silently
  }

  // Final results (relative to `cwd`) and indicator if we reached the limit
  const results: string[] = []

  // Track directories we've seen to avoid duplicate "(files omitted)" markers
  const processedDirs = new Set<string>()

  // BFS queue
  // Each entry holds the directory path, patterns, and relative path
  const queue: Array<{ path: string; patterns: string[]; relPath: string }> = [
    {
      path: resolve(dirPath),
      patterns: rootPatterns,
      relPath: relative(cwd, resolve(dirPath)).replace(/\\/g, '/') || '.',
    },
  ]

  // Perform BFS until queue is empty or maxCount is reached
  while (queue.length > 0) {
    // biome-ignore lint/style/noNonNullAssertion: checked above
    const { path: currentPath, patterns: parentPatterns, relPath: currentRelPath } = queue.shift()!

    // Mark this directory as processed
    processedDirs.add(currentRelPath)

    // Merge parent's patterns with local .gitignore
    const mergedPatterns = await extendPatterns(parentPatterns, currentPath)
    const folderIg = createIgnore(mergedPatterns)

    const entries = await fs.readdir(currentPath, { withFileTypes: true })
    entries.sort((a, b) => a.name.localeCompare(b.name)) // Sort entries for consistent order

    for (const entry of entries) {
      const fullPath = join(currentPath, entry.name)
      // Convert full path to something relative to `cwd`
      const relPath = relative(cwd, fullPath).replace(/\\/g, '/')

      if (folderIg.ignores(relPath)) {
        continue // Skip ignored entries
      }

      if (entry.isDirectory()) {
        if (recursive) {
          queue.push({
            path: fullPath,
            patterns: mergedPatterns,
            relPath,
          })
        }
      } else {
        results.push(relPath)
        if (results.length >= maxCount) {
          // We've hit the limit, add "(files omitted)" markers for directories
          // still in the queue and the current directory if we haven't processed all its files

          // First, check if there are remaining files in the current directory
          const remainingEntries = entries.slice(entries.indexOf(entry) + 1)
          const hasRemainingFiles = remainingEntries.some(
            (e) => !e.isDirectory() && !folderIg.ignores(relative(cwd, join(currentPath, e.name)).replace(/\\/g, '/')),
          )

          if (hasRemainingFiles) {
            const marker = `${currentRelPath}/(files omitted)`
            results.push(marker)
          }

          // Then add markers for all directories still in the queue
          for (const queueItem of queue) {
            // Only add markers for directories we haven't processed yet
            if (!processedDirs.has(queueItem.relPath)) {
              const marker = `${queueItem.relPath}/(files omitted)`
              results.push(marker)
              processedDirs.add(queueItem.relPath) // Mark as processed to avoid duplicates
            }
          }

          results.sort()
          return [results, true]
        }
      }
    }
  }

  results.sort()
  // If we exhaust the BFS queue, we did not reach maxCount
  return [results, false]
}
