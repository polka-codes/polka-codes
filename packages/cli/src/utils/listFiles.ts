import { promises as fs } from 'node:fs'
import { join, relative, resolve } from 'node:path'
import ignore, { type Ignore } from 'ignore'

/** Default patterns commonly ignored in projects of various languages. */
const DEFAULT_IGNORES = [
  '__pycache__',
  '.DS_Store',
  '.env',
  '.git',
  '.idea',
  '.svn',
  '.temp',
  '.vscode',
  'coverage',
  'dist',
  'node_modules',
  'out',
  'Thumbs.db',
]

/**
 * Reads a `.gitignore` file in `dirPath` (if it exists) and appends its lines
 * to the `basePatterns`. Returns a new array without mutating the original.
 */
async function extendPatterns(basePatterns: string[], dirPath: string): Promise<string[]> {
  try {
    const gitignorePath = join(dirPath, '.gitignore')
    const content = await fs.readFile(gitignorePath, 'utf8')
    const lines = content.split(/\r?\n/).filter(Boolean)
    return [...basePatterns, ...lines]
  } catch {
    // No .gitignore or unreadable
    return basePatterns
  }
}

/** Creates an `ignore` instance from the given patterns. */
function createIgnore(patterns: string[]): Ignore {
  return ignore().add(patterns)
}

/**
 * Lists files under `dirPath` in BFS order, respecting:
 *   - A default set of ignores
 *   - A root .gitignore under `cwd`
 *   - Any .gitignore files in child directories (merged as we go)
 *
 * Returns `[files, limitReached]`:
 *   - `files` is the array of file paths (relative to `cwd`)
 *   - `limitReached` is `true` if `maxCount` was hit, otherwise `false`
 */
export async function listFiles(dirPath: string, recursive: boolean, maxCount: number, cwd: string): Promise<[string[], boolean]> {
  // Merge default ignores with root .gitignore (if found)
  let rootPatterns = [...DEFAULT_IGNORES]
  try {
    const rootGitignore = await fs.readFile(join(cwd, '.gitignore'), 'utf8')
    const lines = rootGitignore.split(/\r?\n/).filter(Boolean)
    rootPatterns = [...rootPatterns, ...lines]
  } catch {
    // No .gitignore at root or unreadable; ignore silently
  }

  // Final results (relative to `cwd`) and indicator if we reached the limit
  const results: string[] = []

  // BFS queue
  // Each entry holds the directory path and the patterns inherited from its parent
  const queue: Array<{ path: string; patterns: string[] }> = [{ path: resolve(dirPath), patterns: rootPatterns }]

  // Perform BFS until queue is empty or maxCount is reached
  while (queue.length > 0) {
    // biome-ignore lint/style/noNonNullAssertion: checked above
    const { path: currentPath, patterns: parentPatterns } = queue.shift()!

    // Merge parent's patterns with local .gitignore
    const mergedPatterns = await extendPatterns(parentPatterns, currentPath)
    const folderIg = createIgnore(mergedPatterns)

    const entries = await fs.readdir(currentPath, { withFileTypes: true })

    for (const entry of entries) {
      const fullPath = join(currentPath, entry.name)
      // Convert full path to something relative to `cwd`
      const relPath = relative(cwd, fullPath).replace(/\\/g, '/')

      if (folderIg.ignores(relPath)) {
        continue // Skip ignored entries
      }

      if (entry.isDirectory()) {
        if (recursive) {
          queue.push({ path: fullPath, patterns: mergedPatterns })
        }
      } else {
        results.push(relPath)
        if (results.length >= maxCount) {
          results.sort()
          // Stop searching as soon as we reach maxCount
          return [results, true]
        }
      }
    }
  }

  results.sort()
  // If we exhaust the BFS queue, we did not reach maxCount
  return [results, false]
}
