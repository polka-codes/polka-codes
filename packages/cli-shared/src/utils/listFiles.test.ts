import { afterAll, beforeAll, describe, expect, it } from 'bun:test'
import { promises as fs } from 'node:fs'
import { join } from 'node:path'

import { listFiles } from './listFiles'

describe('listFiles', () => {
  const testDir = join(__dirname, 'test-fixtures')

  beforeAll(async () => {
    // Create test directory structure
    await fs.mkdir(testDir, { recursive: true })
    await fs.writeFile(join(testDir, '.gitignore'), 'ignored.txt\nsubdir/ignored-too.txt')
    await fs.writeFile(join(testDir, 'file1.txt'), '')
    await fs.writeFile(join(testDir, 'file2.txt'), '')
    await fs.writeFile(join(testDir, 'ignored.txt'), '')

    // Create subdirectory
    const subDir = join(testDir, 'asubdir')
    await fs.mkdir(subDir)
    await fs.writeFile(join(subDir, 'file3.txt'), '')
    await fs.writeFile(join(subDir, 'ignored-too.txt'), '')
  })

  afterAll(async () => {
    // Clean up test directory
    await fs.rm(testDir, { recursive: true, force: true })
  })

  it('should list files in directory', async () => {
    const [files] = await listFiles(testDir, false, 10, testDir, [])
    expect(files).toEqual(['.gitignore', 'file1.txt', 'file2.txt'])
  })

  it('should list files recursively', async () => {
    const [files] = await listFiles(testDir, true, 10, testDir)
    expect(files).toMatchSnapshot()
  })

  it('should respect maxCount and show truncation markers', async () => {
    const [files, limitReached] = await listFiles(testDir, true, 3, testDir)
    expect(limitReached).toBe(true)

    expect(files).toMatchSnapshot()
  })

  it('should show truncation markers for root directories', async () => {
    // Set a low maxCount to force truncation
    const [files, limitReached] = await listFiles(testDir, true, 2, testDir)
    expect(limitReached).toBe(true)

    expect(files).toMatchSnapshot()
  })

  it('should not show truncation markers when limit not reached', async () => {
    // Set a high maxCount to avoid truncation
    const [files, limitReached] = await listFiles(testDir, true, 100, testDir)
    expect(limitReached).toBe(false)

    expect(files).toMatchSnapshot()
  })

  it('should handle empty directory', async () => {
    const emptyDir = join(testDir, 'empty')
    await fs.mkdir(emptyDir)
    const [files] = await listFiles(emptyDir, true, 10, emptyDir)
    expect(files).toEqual([])
    await fs.rm(emptyDir, { recursive: true })
  })

  it('should handle non-existent directory', async () => {
    expect(listFiles(join(testDir, 'nonexistent'), false, 10, testDir)).rejects.toThrow()
  })

  it('should exclude files matching excludeFiles patterns', async () => {
    const [files] = await listFiles(testDir, true, 10, testDir, ['file2.txt', 'asubdir/*'])
    expect(files).toEqual(['.gitignore', 'file1.txt'])
  })
})
