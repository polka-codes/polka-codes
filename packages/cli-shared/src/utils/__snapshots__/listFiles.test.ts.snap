// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`listFiles should list files recursively 1`] = `
[
  ".gitignore",
  "asubdir/file3.txt",
  "asubdir/ignored-too.txt",
  "file1.txt",
  "file2.txt",
]
`;

exports[`listFiles should respect maxCount and show truncation markers 1`] = `
[
  ".gitignore",
  "asubdir/(files omitted)",
  "file1.txt",
  "file2.txt",
]
`;

exports[`listFiles should show truncation markers for root directories 1`] = `
[
  "./(files omitted)",
  ".gitignore",
  "asubdir/(files omitted)",
  "file1.txt",
]
`;

exports[`listFiles should not show truncation markers when limit not reached 1`] = `
[
  ".gitignore",
  "asubdir/file3.txt",
  "asubdir/ignored-too.txt",
  "file1.txt",
  "file2.txt",
]
`;
